pkg/operator/k8sutil/events.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package k8sutil for Kubernetes helpers.
+package k8sutil
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+)
+
+type eventObject struct {
+	eventsReportedAt time.Time
+	eventCount       int
+}
+
+// EventReporter is custom events reporter type which allows user to limit the events
+type EventReporter struct {
+	recorder       record.EventRecorder
+	reportedEvents map[string]*eventObject
+
+	// report events x times where x is count
+	count int
+
+	// report events after x minutes
+	eventReportAfterMinutes int
+}
+
+// NewEventReporter returns EventReporter object
+func NewEventReporter(recorder record.EventRecorder, maxCountInGivenTime, reportAfter int) *EventReporter {
+	er := &EventReporter{
+		recorder:                recorder,
+		count:                   maxCountInGivenTime,
+		eventReportAfterMinutes: reportAfter,
+	}
+
+	er.reportedEvents = map[string]*eventObject{}
+
+	return er
+}
+
+// Report records a events if eventReportAfterMinutes has passed or events occurred less than count
+func (rep *EventReporter) Report(instance runtime.Object, eventType, eventReason, msg string) {
+
+	eventKey, err := getEventKey(instance, eventType, eventReason, msg)
+	if err != nil {
+		return
+	}
+
+	eventobj, ok := rep.reportedEvents[eventKey]
+	if !ok {
+		logger.Info("Reporting Event ", eventKey)
+		eventobj = &eventObject{eventsReportedAt: time.Now(), eventCount: 1}
+		rep.reportedEvents[eventKey] = eventobj
+		rep.recorder.Event(instance, eventType, eventReason, msg)
+	} else if eventobj.eventsReportedAt.Add(time.Minute * time.Duration(rep.eventReportAfterMinutes)).Before(time.Now()) {
```

Comment:
Can you add some comments to each of these reasons where events are being reported? Actually I don't understand why we reset the counter to 1 after the time elapses. Shouldn't we just keep increasing the count?

Reply:
Done

Reply:
Can you explain more why we need to rate limit? It seems like we should either report an event once, or every time it happens. 

Or perhaps we could use a different approach such as when an event is raised, do the following:
1. If the first event (since the operator was last restarted), report the event
2. If there was an event previously reported, if the new event is the same, do nothing
3. If there was an event previously reported and the event is different, report the new event

With that approach, you would only see a new event whenever there is some status that changed, which seems ideal.

Reply:
This can be done via giving count as 1 and time as 60 min. Because An Event will stay for 1 hour in the k8s cluster.

Reply:
@iamniting Count 1 and time 60 minutes makes sense for that behavior, just one question... What happens if we have an error, then success, then the same error again (all in less than 60 min)? If there is an error after success, we would want it to be reported again.

Reply:
In that case, it won't report the same error as the event again.

Reply:
@travisn I have added support for the same. It won't report the event if it is the same as the last one. It will report same event only if 60 minutes have passed.

Reply:
Sorry missed this thread earlier, events are expensive on etcd. Each on is an API object. If we did not rate limit,  and the report function is called repeatedly at every failed reconcile as the reconciler loops until it succeeds.

Reply:
here is an example:

```
Events:
  Type     Reason              Age    From               Message
  ----     ------              ----   ----               -------
  Warning  ReconcileFailed     2m43s  ClusterController  failed to reconcile cluster "my-cluster": failed to configure local ceph cluster: failed to perform validation before cluster creation: mon count 4 cannot be even, must be odd to support a healthy quorum
  Warning  ReconcileFailed     79s    ClusterController  failed to reconcile cluster "my-cluster": failed to configure local ceph cluster: failed to create cluster: failed to start ceph monitors: failed to assign pods to mons: CANCELLING CURRENT ORCHESTRATION
  Normal   ReconcileSucceeded  39s    ClusterController  cluster has been configured successfully
```

```
Events:
  Type     Reason              Age                 From               Message
  ----     ------              ----                ----               -------
  Warning  ReconcileFailed     109s                ClusterController  failed to reconcile cluster "my-cluster": failed to configure local ceph cluster: failed to create cluster: failed to start ceph monitors: failed to assign pods to mons: CANCELLING CURRENT ORCHESTRATION
  Normal   ReconcileSucceeded  69s                 ClusterController  cluster has been configured successfully
  Warning  ReconcileFailed     2s (x2 over 3m13s)  ClusterController  failed to reconcile cluster "my-cluster": failed to configure local ceph cluster: failed to perform validation before cluster creation: mon count 4 cannot be even, must be odd to support a healthy quorum
```

Reply:
Sounds good, I like the new approach of only reporting the event if it is different from the previously reported event.